### PR TITLE
lib/select: Port Cockpit selects to native HTML

### PR DESCRIPTION
**Pull request by garrett**
TODO:
- [x] Finish HTML and adjust tests (pitti)
- [x] Port CSS over from my [inital demo](https://codepen.io/garrett/pen/jdEoxe) and make it actually pixel-perfect with our existing dropdowns (this is for me to do)
- [x] Integrate ported CSS and adjust if necessary
- [x] Fix nesting of optgroup (pitti)
- [x] Fix too narrow "Create RAID" selectors (garrett)
- [x] Check if the above fix also covers "Create Partition Table" and "Create Partition" dialog selectors (same issue)
- [x] Fix too narrow "Graphics Console" selector
- [x] Fix too narrow unit selector in Machines "Add Disk"
- [x] Re-drop the "on <weekday>" and selector change commits, they are intended for a follow-up
- [x] Test on Firefox
- [x] Test on Chromium
- [x] Test on iOS (lars)
- [x] Test on Android Firefox (pitti)
- [x] Test on Android Chromium

### CODE REVIEW

**Comment by garrett** [comment]
This is a first rough attempt at porting our custom dropdowns to native selects.

TODO:
- [ ] Finish HTML (I need your help with this)
- [ ] Port CSS over from my [inital demo](https://codepen.io/garrett/pen/jdEoxe) (this is for me to do)
- [ ] Test, test, test (this is something for all of us)

**Comment by garrett** [comment]
Reasoning: Our custom dropdowns are not accessible... at all. Swapping them out for styled native dropdowns would fix a11y.

**Comment by martinpitt** [comment]
I force-pushed a rebase to current master, as this PR broke our CI (it was based off  a too old master which didn't yet know about `bots/tests-invoke --pull-number`).

File: pkg/lib/cockpit-components-select.jsx [hunk 1]
```diff
@@ -192,26 +166,30 @@ export class SelectEntry extends React.Component {
...
 
 /* Header
  * Example: <SelectHeader>Some header</SelectHeader>
  */
 export const SelectHeader = ({ children }) => {
     const value = (children !== undefined) ? children : textForUndefined;
+    /* HACK: The optgroup should (optionally) wrap options */
     return (
-        <li className="dropdown-header">{value}</li>
+        <optgroup label={value} />
```
[the remark this answers is not in the document]
Comment (garrett) [reply pkg/lib/cockpit-components-select.jsx]: Oh, I had a comment from our hackfest to show where I need help. Oooooops. GitHub did save it at least, in a pending state. :wink:

**Comment by garrett** [comment]
I've forked and updated the Codepen to have CSS styling that matches PatternFly:

Codepen demo @ https://codepen.io/garrett/pen/exjrJB

screenshot_2019-02-13 html select testing the limits 1

Reference: https://www.patternfly.org/pattern-library/widgets/#bootstrap-select

screenshot_2019-02-13 widgets patternfly 2


Seeing it side-by-side like this makes me realize it's off by a few pixels. Oops. But I think it might be related to being implemented outside of Cockpit & PatternFly.

It's probably close enough to start integrating the CSS and fine-tuning once it's in Cockpit within this PR.

**Comment by martinpitt** [comment]
I force-pushed an initial update of this. This should by and large work functionally, there's just no CSS adjustment at all yet. Hence I unticked the checkbox above. I fixed one case of now-wrong custom CSS and ported one test case, but there are a lot more which need porting still. I want to see one full test run to get a complete todo list.

**Comment by martinpitt** [comment]
What I mean by "missing CSS": Current master:
auto-updates-old

With this PR (without CSS adjustments), most selects now look like this:

auto-updates-new

The storage page seems to have some custom CSS. Dialog in master:

storage-old

and with this PR:

storage-new

File: pkg/lib/cockpit-components-select.jsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (martinpitt) [reply pkg/lib/cockpit-components-select.jsx]: With the current structure of the components we can't properly group them. It seems to work as a single element too, though? If not (in e. g. some browsers), we have to restructure the components to express which items are grouped.

**Comment by garrett** [comment]
I have CSS added locally, and it's kind of mostly working.

HOWEVER, CSP gets in the way, as I tried to use inline SVGs:
- Simple info about CSP and SVGs: https://stackoverflow.com/questions/18447970/content-security-policy-data-not-working-for-base64-images-in-chrome-28#18449556
- Details: https://security.stackexchange.com/questions/94993/is-including-the-data-scheme-in-your-content-security-policy-safe/95011#95011

(The error message is `Content Security Policy: The page’s settings blocked the loading of a resource at inline (“default-src”).`)

This means, we'll either need to relax the CSP or store the small SVG somewhere and load it like any other image. I'd opt for the second of these two options.

So: Where should we put an image that may be loaded by all of Cockpit?

There's a server-small.png we load from `/pkg/shell/images/` — but that feels wrong. Aside from favicon.ico, that's the only graphic we load though. Is there a better place?

**Comment by garrett** [comment]
Oddly, GitHub doesn't support SVG files... but does support ZIP files. So here's the teeny-tiny SVG in a ZIP.

[dropdown-caret.zip](https://github.com/cockpit-project/cockpit/files/2905880/dropdown-caret.zip)

**Comment by martinpitt** [comment]
At the risk of repeating an IRC discussion: Using the [fa-angle-down icon](https://www.patternfly.org/styles/icons/) isn't good enough?

**Comment by garrett** [comment]
I mentioned in IRC and during today's meeting that we cannot use fontawesome, as it's a glyph. Native dropdowns sadly do not support `:before` or `:after` with a high-numbered UTF-8 glyph as generated text content, which is the way you use an icon font.

To style the dropdown, it needs a background (and in our case, two background images: a generated gradient background and the SVG caret glyph as SVG).

So, yes, we do need to be able to put a graphic _somewhere_ that is available to all of Cockpit and reference it in CSS. :wink:

(Unless we weaken CSP for data: URIs. But it's better to simply put the graphic somewhere and source it via CSS.)

**Comment by martinpitt** [comment]
Then I think we should ship it in /usr/share/cockpit/static/ . I can look into how to get it there (for crafting the CSS you can just copy it there manually locally)

**Comment by martinpitt** [comment]
The downside of shipping it centrally (in -bridge or -ws) is that this would break running older cockpit-ws against newer cockpit pages. This can happen with mixed versions on RHEL/CentOS 7, or when using remote machines. So this is too brittle, let's not ship it centrally, but per-page (or, if at all possible, avoid it entirely and use the standard select's caret. It's slightly stronger, I know, but maybe that's still acceptable)

**Comment by garrett** [comment]
PatternFly 4 uses `data:` URIs for background images in CSS all over the place, including [select dropdowns](https://pf4.patternfly.org/components/FormControl/examples/) as well as error states.

So this issue will hit us again in the near future.

Is it possible to have a bugfix backport patch to adjust the CSP for all supported versions of Cockpit (such as RHEL 7.6)? Otherwise, I don't think we'll ever be able to move to PF4. :cry:

**Comment by martinpitt** [comment]
Current plan after long IRC discussion:
 - Use inline SVG in the CSS. Going forward this is the much more flexible plan than shipping static files in -ws (which are really hard to rely on)
 - Add `img-src data:` to `cockpit_web_response_security_policy()` now, it should be fine security-wise.
 - We can't rely on that yet (older ws/bridge, remote machines, etc.), so add it to our manifests as well for the time being.
 - In a few months or so, when we don't support platforms without the new policy any more, we can drop the CSP from the manifests again.

@larskarlitski, @garrett, does that work for you?

**Comment by larskarlitski** [comment]
> @larskarlitski, @garrett, does that work for you?

Perfect. Thanks!

**Comment by martinpitt** [comment]
> Add `img-src data:` to `cockpit_web_response_security_policy()` now

I sent PR #11268 to do that.

**Comment by martinpitt** [comment]
@garrett : FYI, I did another force-push to fix more tests (still not done yet), plus rebasing to current master. Do you want me to already add a commit that extends C-S-P to allow `img-src data:`?

**Comment by martinpitt** [comment]
I did a new push. This should fix the last remaining failure on fedora-29, so triggering all tests now.

> Do you want me to already add a commit that extends C-S-P to allow img-src data:?

@garrett: This is now done (separate commit here). Thus your inline CSS SVG should work now, please let me know if it doesn't.

**Comment by martinpitt** [comment]
Some more tests to fix:
 - [ ] https://fedorapeople.org/groups/cockpit/logs/pull-11155-20190228-141515-e8adafa0-verify-rhel-atomic/log.html#191
 - [ ] https://fedorapeople.org/groups/cockpit/logs/pull-11155-20190228-141515-e8adafa0-verify-debian-testing/log.html#169
 - [ ] https://fedorapeople.org/groups/cockpit/logs/pull-11155-20190228-141515-e8adafa0-selenium-chrome/log.html#30

**Comment by martinpitt** [comment]
Next round. This should now fix all tests (famous last words).

**Comment by martinpitt** [comment]
Ah, must adjust the sub-man test for [7-6-distropkg](https://fedorapeople.org/groups/cockpit/logs/pull-11155-20190301-125353-21f218ec-verify-rhel-7-6-distropkg/log.html#264)

**Comment by martinpitt** [comment]
This still breaks one [edge selenium test](https://fedorapeople.org/groups/cockpit/logs/pull-11155-20190302-083131-28eb7b6c-selenium-edge/log.html#29).

**Comment by martinpitt** [comment]
I found a workaround for the Edge test bug, pushed. :crossed_fingers:

**Comment by martinpitt** [comment]
Yay, all tests green! So I take the liberty to tick "finish html and ajust tests" as "done". Over to @garrett, for adding CSS magic? (This should work now, see above)

**Comment by garrett** [comment]
Thanks! Nice!

**Comment by garrett** [comment]
CSS that depends on `.dropdown` that needs to be fixed (or at least inspected) for `.ct-dropdown` too:

```
pkg/shell/shell.less:147:#dashboard-toolbar > div.dropdown {

pkg/machines/machines.less:230:table.vcpu-detail-modal-right .btn-group.bootstrap-select.dropdown {

pkg/networkmanager/networking.css:62:#networking-graph-toolbar .dropdown {
pkg/networkmanager/networking.css:66:#network-interface-graph-toolbar .dropdown button span,
pkg/networkmanager/networking.css:67:#networking-graph-toolbar .dropdown button span {

pkg/lib/form-layout.less:177:    &:not(.ct-form-layout-relax) > .dropdown {

pkg/systemd/services.css:47:    .filter-group .dropdown {

pkg/systemd/host.css:146:#server-graph-toolbar .dropdown {
pkg/systemd/host.css:251:#shutdown-dialog .dropdown {

pkg/packagekit/updates.less:340:    .dropdown {
pkg/packagekit/updates.less:348:.auto-conf-group:first-child > .dropdown {

pkg/docker/docker.css:307:.has-error > .dropdown > button > span {

pkg/storaged/storage.css:338:#storage-graph-toolbar .dropdown {
pkg/storaged/storage.css:342:#storage-graph-toolbar .dropdown button span {
```

Easiest approach would be to duplicate the CSS line for each (and switch to a comma) and sub out `.dropdown` for `.ct-dropdown` in each line... then it should support both the dropdown class (like from PatternFly) as well as our custom dropdowns. In the cases where multiple types of dropdowns on the page are _not_ used, we can just sub it out without duplicating the line.

I'm guessing most of these mainly add margins and/or adjust widths.

**Comment by garrett** [comment]
I've added some CSS that works pretty well... and was working on browser testing last night when my machine crashed hard. As it was already getting late, I just pushed what I had this morning (after rebasing & squashing a little bit).

Highlights:

- Styled like our non-native dropdowns!
- Adapted the parts of Cockpit that use `.dropdown` in the CSS for spacing & width to use `.ct-dropdown` too
- Better divider
- Demo has divider now too

**Comment by garrett** [comment]
I need to continue browser testing and adapting the CSS.

I was also working on making a more pleasant focus ring, but it's tricky. (The default is not too horrible — it's just not quite as nice — I might just go back to that.)

But I think it's time we can start testing it in various places in various browsers.

File: pkg/lib/cockpit-components-select.jsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (garrett) [reply pkg/lib/cockpit-components-select.jsx]: It doesn't work as a single element like this. It only works as an open, unterminated tag or as a tag pair that wraps options.

The way I have it implemented here is that it opens and closes the group immediately, which is wrong.

We need either:

1. `<optgroup><option></optgroup>`
2. `<optgroup><option>`

But the code is effectively doing:

- `<optgroup></optgroup><option>`

The problem is: I don't know how to start a tag like this in React without closing it. (I think I got an error when I didn't terminate it?)

**Comment by garrett** [comment]
I forgot to follow-up on the optgroup issue earlier. Here's my reply: https://github.com/cockpit-project/cockpit/pull/11155#discussion_r262426963

We don't use select grouping in many places, but I think we do use it in the new VM dialog.

**Comment by martinpitt** [comment]
The selects work a lot better now, thanks @garrett! 

All of the selects now use a bold-ish serif font, which looks out of place. In addition, the "Create RAID" dialog doesn't fill up the entire width any more. 
create-raid

Comparison with current master:

storage-old

**Comment by garrett** [comment]
I've added a hotfix for the font issue @ https://github.com/cockpit-project/cockpit/pull/11155/commits/fbd1642fe0dc8f13359c75362616c614201d62e7 — intended to be merged in with the HTML+CSS commits (when that happens). Do a pull to get the fix.

I'll look into the width thing now. It's probably something I'll need to fix in `ct-form-layout`, I'm guessing.

**Comment by martinpitt** [comment]
I pushed a fixup for the optgroup nesting. (should also be squashed eventually)

**Comment by martinpitt** [comment]
I went through all usages of the Select component, and found a few more. The "Create Partition" and "Create Partition Table" ones look very similar to the above "Create RAID", and are probably covered by the same fix:

create-parttable
create-part

Debugging tip: To get a scratch RAM drive, use `sudo modprobe  scsi_debug dev_size_mb=300`.

The unit selector on Machines "Add disk" is too narrow:

disks-add

And finally, the Machine console selector is too narrow again, despite my first patch on this PR:

graphics-console

I added TODO items to the description for all of these, and also one for re-dropping the accidental "sw updates" changes.

**Comment by martinpitt** [comment]
One new test failure, which probably came out of the fixups: https://fedorapeople.org/groups/cockpit/logs/pull-11155-20190305-114242-77acf6fd-verify-rhel-atomic/log.html#194

Unrelated flake: https://fedorapeople.org/groups/cockpit/logs/pull-11155-20190305-114242-77acf6fd-verify-fedora-i386/log.html#223

**Comment by andreasn** [comment]
It seems ct-select has a width of 220px, that's probably why they end up so narrow.

**Comment by martinpitt** [comment]
I re-tested all the selects, they look great now. I squashed the commits, and took  out the PackageKit ones. From my POV this is fine now and ready to land.

**Comment by martinpitt** [comment]
I also tested everything in Chromium now. The select dropdowns have light-gray background instead of white (as in Firefox), but it looks and works consistently as well. If you want to change this at all, it would IMHO be okay for a follow-up, but I'll leave that up to you.

dropdown-chromium

**Comment by garrett** [comment]
"The select dropdowns have light-gray background instead of white (as in Firefox), but it looks and works consistently as well." ← This is styled natively and cannot be changed by CSS, except (in a very limited capacity) in Chrome. As I'm pretty sure Chrome would allow it, I can look into changing it to be a white background for the dropdowns too. :+1:

**Comment by garrett** [comment]
> width of 220px

I was copying our old dropdowns. I removed it yesterday in my big batch of changes and pushed it up last night. It's fine everywhere I checked, especially since we're using `ct-form-layout` in a lot of places now (and therefore don't really need the hardcoded width by default).

**Comment by garrett** [comment]
Force-pushed with a fix for white backgrounds on options, supported only in Chrome/Chromium (AFAICT). It's @ https://github.com/cockpit-project/cockpit/pull/11155/files#diff-b7bce1a3b916cdaba71d9fc692584e55R280 in case you're wondering what it looks like. (Pretty simple, really.)

**Comment by martinpitt** [comment]
I tested mobile Firefox (actually, Fennec) on Android. The selects work/look fine. Just two  observations:

 - Some of them use in-page dropdowns, some of them the browser native "modal OS window with the options".
 - The divider option (-------) appears in  the above popup. It's (correctly) unselectable, just looks a wee bit odd. But IMHO that's fodder for a follow-up, it doesn't really break anything.

**Comment by garrett** [comment]
I tested in Firefox on Android and it's fine too. I also tested in Chrome (actually Bromite, the FOSS, deGoogled fork of Chrome) and it's basically the same too (which is good in this case).

1. In-page dropdowns in Cockpit are done by either a custom copy-pasted widget and/or PatternFly... and not our custom dropdown (which is now a styled native select in this PR).
2. There's nothing we can do about the divider. This is the best it gets. That's just how it is.

Oh... We probably want to ARIA label it so it doesn't say "**box drawing line box drawing line box drawing line box drawing line box drawing line box drawing line**" and so on.

**Comment by garrett** [comment]
I've looked everywhere for accessibility-related information on selects, and specifically options and screen reader usage. I can't find a solution.

As the item is disabled and has a role assigned to "divider", it should work. Probably. But who knows what different screen readers on all the different platforms with different browsers actually do?

I tried Firefox devtools accessibility tab and looked at the accessibility tree... and see the role is supported there. And aria-label doesn't seem to work, nor does the label attribute.

In VoiceOver on my Mac using Safari, I get "———————, dimmed" shown on the screen and it just says "comma dimmed". I guess that's OK?

As it seems label and aria-label have no effect on the `<option>` element, and role seems to (despite docs saying that it doesn't support roles, oddly) and at least VoiceOver does a mostly-right thing, and this is overwhelmingly much improved overall for accessibility with regard to screen reader usage, keyboard navigation, and mobile friendliness (with larger tap areas), _I'm considering this done_ and any theoretical screen reader improvements would go into another PR.

**Pull request merged**